keras_spiking/constraints.py

```diff
@@ -0,0 +1,31 @@
+import tensorflow as tf
+
+
+class Mean(tf.keras.constraints.Constraint):
+    """Constrains weight tensors to be their mean.
+
+    Parameters
+    ----------
+    axis : int
+        Axis used to compute the mean and repeat its value. Defaults to the last axis.
+    non_neg : boolean
+        Whether to constrain the resulting mean to be non-negative as well.
+    """
+
+    def __init__(self, axis=-1, non_neg=False):
+        self.axis = axis
+        self.non_neg = non_neg
+        self._post_constraint = tf.keras.constraints.NonNeg() if non_neg else None
+
+    def __call__(self, w):
+        r = tf.repeat(
+            input=tf.math.reduce_mean(w, axis=self.axis, keepdims=True),
+            repeats=tf.shape(w)[self.axis],
```

Comment:
Do you know if this works if the shape for that dimension is `None`?

Reply:
I have not tried.

Reply:
I'm trying to think of a way to test this, but I'm not sure how to get into a situation where a weight tensor has a `None` dimension? Since this is always going to be passed a weight tensor via `Layer.add_weight` is it possible for this to happen?

Reply:
Yeah, so I guess the first question is does `Lowpass` even support having `None` for the size. From what I can tell from this simple script, it doesn't. This is actually something we might want a better error for, because right now it's a bit hard to parse (the error results in another error; the first error hints at the problem, but the second really obscures things). 
```
inp = x = tf.keras.Input([None, None])
x = keras_spiking.Lowpass(tau=0.005)(x)
model = tf.keras.Model(inp, [x])
y = model.predict(np.ones(1, 20, 5))
```
So in short, don't worry about supporting `None` here!

Reply:
I also tried the same thing with `Dense` and got the error: `ValueError: The last dimension of the inputs to `Dense` should be defined. Found `None`.` so this problem might not be unique to lowpass either. But it sounds like this could have a safety check to fail faster with a better error message (for dense it happens inside of `build` as early as possible).

Reply:
`tf.shape(w)` will resolve the dynamic shape of `w` at runtime, so it would never be `None` (even if you could create a weight with an unspecified shape at build time, which as Aaron says I don't think is possible). So this should be fine :+1: